### PR TITLE
Feat/add lock timeout option

### PR DESCRIPTION
**Pull request by emalihin**

### CODE REVIEW

File: api/v1alpha1/terraform_types.go [hunk 1]
```diff
@@ -120,6 +120,10 @@ type TerraformSpec struct {
 	// +required
 	Interval metav1.Duration `json:"interval"`
 
+	// Duration to retry a state lock.
+	// +optional
+	LockTimeout string `json:"lockTimeout,omitempty"`
+
```
Comment (Nalum) [inline api/v1alpha1/terraform_types.go:126]: Should this be part of the `TFStateSpec`?

**Review by Nalum — CHANGES_REQUESTED** [review]
Looks good, just one comment on the location of the field. @chanwit thoughts?

File: api/v1alpha1/terraform_types.go (hunk 1 above)
Comment (chanwit) [reply api/v1alpha1/terraform_types.go:126]: Also this duration would be of the Duration type (from the metav1 package).

Comment (emalihin) [reply api/v1alpha1/terraform_types.go]: thanks for the feedback, wasn't sure about these two 👌 changed

File: controllers/tf_controller_plan.go [hunk 2]
```diff
@@ -59,6 +59,12 @@ func (r *TerraformReconciler) plan(ctx context.Context, terraform infrav1.Terraf
 		planRequest.Destroy = true
 	}
 
+	// check if Lock Timeout is set
+	if terraform.Spec.LockTimeout != "0s" {
+		log.Info("lock timeout is set")
+		planRequest.LockTimeout = terraform.Spec.LockTimeout
```
Comment (Nalum) [inline controllers/tf_controller_plan.go:65]: Don't forget to update this :slightly_smiling_face:

**Review by Nalum — CHANGES_REQUESTED** [review]
Looks good, just forgot to update the plan call.

If you could look at adding a test that would also be great!

File: controllers/tf_controller_plan.go (hunk 2 above)
Comment (emalihin) [reply controllers/tf_controller_plan.go]: trying to wrap my head around how testing works :)

File: controllers/tf_controller_plan.go [hunk 3]
```diff
@@ -59,6 +59,12 @@ func (r *TerraformReconciler) plan(ctx context.Context, terraform infrav1.Terraf
 		planRequest.Destroy = true
 	}
 
+	// check if Lock Timeout is set
+	if terraform.Spec.TFState.LockTimeout.Duration.String() != "" {
```
Comment (chanwit) [inline controllers/tf_controller_plan.go:63]: We got nil reference here in the test: 
https://github.com/weaveworks/tf-controller/actions/runs/3854570045/jobs/6699132334

Could you please add an if to check that TFState is not nil here?

The frame:
```
E0114 13:17:24.579832   18289 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 608 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2b654c0?, 0x423b820})
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:75 +0x99
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:110 +0xb9
panic({0x2b654c0, 0x423b820})
	/opt/hostedtoolcache/go/1.19.4/x64/src/runtime/panic.go:884 +0x212
github.com/weaveworks/tf-controller/controllers.(*TerraformReconciler).plan(_, {_, _}, {{{0x29439c9, 0x9}, {0xc000a46880, 0x20}}, {{0xc0010c4210, 0x26}, {0x0, ...}, ...}, ...}, ...)
	/home/runner/work/tf-controller/tf-controller/controllers/tf_controller_plan.go:63 +0x417
github.com/weaveworks/tf-controller/controllers.(*TerraformReconciler).reconcile(_, {_, _}, {_, _}, {{{0x29439c9, 0x9}, {0xc000a46880, 0x20}}, {{0xc0010c4210, ...}, ...}, ...}, ...)
```

**Review by chanwit — CHANGES_REQUESTED** [review]

File: controllers/tf_controller_plan.go (hunk 3 above)
Comment (emalihin) [reply controllers/tf_controller_plan.go]: thank you for flagging this @chanwit , will my last push fix it?
apologies for the the lack of attention on this PR too

**Comment by Nalum** [comment]
Hey @emalihin, apologies for the delay on getting back to you. I think this is looking good, would you be able to update the branch to bring it in line with the latest changes and we can rerun CI on that?

If you don't think you'll have the time I will pick this up later this week or next week.

**Comment by emalihin** [comment]
> Hey @emalihin, apologies for the delay on getting back to you. I think this is looking good, would you be able to update the branch to bring it in line with the latest changes and we can rerun CI on that?
> 
> If you don't think you'll have the time I will pick this up later this week or next week.

Hi @Nalum - I'd appreciate if you could take over this MR as I'm deep into other work atm.

**Comment by Nalum** [comment]
Closing this in favour of #1087

**Pull request closed without merging**